test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h

```diff
@@ -229,8 +229,9 @@ fuzzExtProcRun(const test::extensions::filters::http::ext_proc::ExtProcGrpcTestC
   FuzzedDataProvider downstream_provider(
       reinterpret_cast<const uint8_t*>(input.downstream_data().data()),
       input.downstream_data().size());
-  FuzzedDataProvider ext_proc_provider(
-      reinterpret_cast<const uint8_t*>(input.ext_proc_data().data()), input.ext_proc_data().size());
+  static std::string ext_proc_data = input.ext_proc_data();
+  static FuzzedDataProvider ext_proc_provider(
```

Comment:
Correct me if I'm wrong, but isn't `ext_proc_data` and `ext_proc_provider` dependent on the fuzz test input?
If so, why should this be static?

Reply:
Thanks for review.

The ext_proc_provider address is passed into the c'tor of a persistent static object: fuzz_helper:

https://github.com/envoyproxy/envoy/blob/73da4c3a3fdd11192ec9f8dc60906e2f3402132d/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L245

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.cc#L106

And the ext_proc_provider is used by the persistent test server here: https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz_helper.cc#L451


Thus ext_proc_provider has to be persistent. And ext_proc_data is a pointer to construct ext_proc_provider so it has to be persistent as well.

In a nutshell, ext_proc_provider is an address used by the persistent fuzz_helper, thus it has to be persistent across different test runs.   Thus the solution is to copy the test input into this static variable and using it to construct fuzz_helper.

Reply:
So I guess either the fuzz_helper object should not be static, or the data should be explicitly updated on each run.

Reply:
Test server is using fuzz_helper to send random response. And we want to have test server persistent to avoid launch it each run.

So, by  copy the test input into this static variable and using it to construct fuzz_helper, the ext_proc server data is updated during each run.

And the downstream_provider, is another part of data derived from the test input. This data is not used to launch Envoy or Test server (both of them are persistent), thus it is non-static:

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L229

https://github.com/envoyproxy/envoy/blob/0bc754a213b6cd28c7117c82de0f721bb19d6ef2/test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.h#L277

Reply:
As discussed offline, I'm still not convinced that the fuzzer will ensure that different iterations will not try to access old/new state objects.
I wonder which static objects make the persistent fuzzer much faster than the non-persistent variant. Can you provide a breakdown of the different objects initialization costs?

Reply:
Sorry for not being able to respond early.
With below code:
  ext_proc_data = std::make_unique<std::string>(input.ext_proc_data());

ext_proc_data is refreshed after each iteration.  I dumped the contents of ext_proc_data and confirmed it matching with input in each iteration.

Regarding the cost of initialize Envoy and test-server,  I did some measurements when we add this persistent fuzzer. I recall it's about ~450ms to start Envoy and ~80ms to start  the test server in my cloud top VM during each iteration.  The persistent fuzzer is just trying to amortize that in 2000 iterations.

Reply:
> ext_proc_data is refreshed after each iteration. I dumped the contents of ext_proc_data and confirmed it matching with input in each iteration.

IIUC this is a flaky test, and so there may be a case where they will not be matched.
I'm going to approve this, but I suggest considering refactoring to use a FakeUpstream, and solve the underlying issue if this persists.